Cabal cleanup feat. new-* commands

fxdpntthm opened the pull request:
from now on you can run:
   - `make alb` to build alb (also pulls in the dependencies)
   - `make ilab` to build ilab (also pulls in the dependencies)
   - `make tests-alb` to run the tests to check sanity of alb compiler
   - `make tests-ilab` to run the tests and check sanity of ilab solver
   - `make install-alb` run tests and dist the alb artifacts
   - `make install-ilab` run tests and dist the ialb artifacts

All the make tasks essentially run cabal new-* commands

Also README styling because its 2019 and markdown looks a little cooler than ascii

fxdpntthm committed:
Cabal cleanup includes the following task

One command to make them all: `make all` should build the project from scratch and potentially run the tests
run tests: Fix the current (non-existent) testing framework by running make tests

Now on you can run:
   - `make alb` to build alb (also pulls in the dependencies)
   - `make ilab` to build ilab (also pulls in the dependencies)
   - `make tests-alb` to run the tests to check sanity of alb compiler
   - `make tests-ilab` to run the tests and check sanity of ilab solver
   - `make install-alb` run tests and dist the alb artifacts
   - `make install-ilab` run tests and dist the ialb artifacts

Minor things picked up because I squashed commits
- Cleaned up Makefile for alb.cabal
- make install as expected

- changed $(BINDIR) in Makefile to be ~/.local/bin
    - Make install will generate a symlink to alb and ilab executables in ~/.local/bin
    - Testing framework seems to work fine now.

- ascii art is cool and so is markdown for readme
- and new installation instructions
- making tests run with `make tests` from tests/Makefile
  only 4 tests run right now.

- Tests now run using cabal new-run alb. No need for bindir unless its installation target.
- make tests-alb runs compiles the files mentioned in tests and then also runs them using runghc testHarness

- this commit fixed `make tests-ilab`
- moved Solver.RunTests into tests as RunTest file should belong to tests and not the src

All the make tasks essentially run cabal new-* commands underneath

TODO!
- Need to figure out why `test/miniprelude.hb fails to parse
- and all the BROKEN_TESTS in `tests/Makefile fail to work
- Fix why -o option ignores the file extension that I provide via comand line.

peschkaj commented:
:+1:

The only error I see happens during tests running when I get:

```
Testing tests/example-test.out: 
testHarness.hs: tests/example-test.out: rawSystem: runInteractiveProcess: exec: does not exist (No such file or directory)
tests/Makefile:108: recipe for target 'tests-alb' failed
make: *** [tests-alb] Error 1
```

peschkaj reviewed (approved):
Looks good 👍

fxdpntthm commented:
So the way the tests run is:

alb first spits out the *.out file using *.hb
testHarness.hs runs it using the process.

The error that you see is because `tests/example-test.out` was not generated 
Is it the case that you see `test/example-test` generated?

Apoorv

> On Feb 10, 2019, at 19:44, Jeremiah Peschka <notifications@github.com> wrote:
> 
> Testing tests/example-test.out: 
> testHarness.hs: tests/example-test.out: rawSystem: runInteractiveProcess: exec: does not exist (No such file or directory)
> tests/Makefile:108: recipe for target 'tests-alb' failed
> make: *** [tests-alb] Error 1
>

peschkaj commented:
Ah, I see why this is failing now: `tests/Makefile` still has `MIL_DIR=$(HOME)/workspace/habit-lang/mil-tools` which is not where mil-tools lives locally. 

The install path for mil-tools is in a `bin` off of whichever directory contains `build.xml`. I suspect on your machine, this is in `~/workspace/habit-lang/mil-tools/`, over here it's in `~/src/habit-lang/mil-tools`. 

Once I change `MIL_DIR` to `../mil-tools` all of the tests to run and pass.

I have pushed that change, but I'd be happy with any better solution.

Commit:
- Assume that mil-tools lives one folder up

- Added a fix in tests/Solver/RunTests.hs to ignore the first line 'Up to date' in the result files.
Manual change is necessary because `cabal new-run ilab` spits an extra line of output. All the tests
will obviously fail if you run `runghc -irc tests/Solver/RunTests.hs` but they will succeed if
you run them again. There is no way to fix this flickering becuase `cabal new-run`'s extra
output cannot be silenced using `-q` option.

fxdpntthm commented on tests/Makefile:
```diff
@@ -90,7 +90,7 @@ ALB_DIR=$(pwd)
 ALB_TEST_DIR=tests
 
 # MIL locations
-MIL_DIR=$(HOME)/workspace/habit-lang/mil-tools
```
This needs a better way..

We can have everyone having different relative paths of their mil-tools and alb compiler
Is there any generic way of solving this problem of relative paths?
